interfaces/builtin/custom_device.go

```diff
@@ -0,0 +1,417 @@
+		return fmt.Errorf(`custom-device %q path must start with / and cannot contain special characters: %q`, attrName, path)
+	}
+
+	if !cleanSubPath(path) {
+		return fmt.Errorf(`custom-device %q path is not clean: %q`, attrName, path)
+	}
+
+	if _, err := utils.NewPathPattern(path); err != nil {
+		return fmt.Errorf(`custom-device %q path cannot be used: %v`, attrName, err)
+	}
+
+	// We don't allow "**" because that's an AppArmor specific globbing pattern
+	// which we don't want to expose in our API contract.
+	if strings.Contains(path, "**") {
+		return fmt.Errorf(`custom-device %q path contains invalid glob pattern "**"`, attrName)
+	}
+
+	return nil
+}
+
+func (iface *customDeviceInterface) validateDevice(path string, attrName string) error {
+	// The device must satisfy udev's device name rules and generic path rules
+	if !customDeviceUDevDeviceRegexp.MatchString(path) {
+		return fmt.Errorf(`custom-device %q path must start with /dev/ and cannot contain special characters: %q`, attrName, path)
+	}
+
+	if err := iface.validateFilePath(path, attrName); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (iface *customDeviceInterface) validatePaths(attrName string, paths []string) error {
+	for _, path := range paths {
+		if err := iface.validateFilePath(path, attrName); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (iface *customDeviceInterface) validateUDevValue(value interface{}) error {
+	stringValue, ok := value.(string)
+	if !ok {
+		return fmt.Errorf(`value "%v" is not a string`, value)
+	}
+
+	if !customDeviceUDevValueRegexp.MatchString(stringValue) {
+		return fmt.Errorf(`value "%v" contains invalid characters`, stringValue)
+	}
+
+	return nil
+}
+
+func (iface *customDeviceInterface) validateUDevValueMap(value interface{}) error {
+	valueMap, ok := value.(map[string]interface{})
+	if !ok {
+		return fmt.Errorf(`value "%v" is not a map`, value)
+	}
+
+	for key, val := range valueMap {
+		if !customDeviceUDevValueRegexp.MatchString(key) {
+			return fmt.Errorf(`key "%v" contains invalid characters`, key)
+		}
+		if err := iface.validateUDevValue(val); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (iface *customDeviceInterface) validateUDevTaggingRule(rule map[string]interface{}, devices []string) error {
+	hasKernelTag := false
+	for key, value := range rule {
+		var err error
+		switch key {
+		case "subsystem":
+			err = iface.validateUDevValue(value)
+		case "kernel":
+			hasKernelTag = true
+			err = iface.validateUDevValue(value)
+			if err == nil {
+				deviceName := value.(string)
+				// furthermore, the kernel name must match the name of one of
+				// the given devices
+				if !strutil.ListContains(devices, "/dev/"+deviceName) {
+					err = fmt.Errorf(`%q does not match a specified device`, deviceName)
+				}
+			}
+		case "attributes", "environment":
+			err = iface.validateUDevValueMap(value)
+		default:
+			err = errors.New(`unknown key`)
```

Comment:
```suggestion
			err = fmt.Errorf(`unknown key %q`, key)
```

Reply:
The actual error returned to the caller is computed a few lines below, and it does contain the key already: that's why I'm not writing it here.

Reply:
But indeed, the error message can be improved: it currently reads

> custom-device "udev-tagging" invalid "foo" tag: unknown key

Which might be confusing, so I'll replace the word "key" with "tag", which hopefully works better.

Reply:
Done.